🌱 hack: implement generator for prowjobs

chrischdi opened the pull request:
<!-- Thanks for sending a pull request! Here are some tips for you:
    1. If this is your first time, please read our contributor guidelines: https://github.com/kubernetes-sigs/cluster-api/blob/main/CONTRIBUTING.md#contributing-a-patch and developer guide https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/book/src/developer/guide.md

    2. Please add an icon to the title of this PR (see https://sigs.k8s.io/cluster-api/CONTRIBUTING.md#contributing-a-patch), and delete this line and similar ones
    the icon will be either ⚠️ (:warning:, major or breaking changes), ✨ (:sparkles:, feature additions), 🐛 (:bug:, patch and bugfixes), 📖 (:book:, documentation or proposals), or 🌱 (:seedling:, minor or other) 
-->

**What this PR does / why we need it**:

Implements `prowjob-gen` a lightweight tool based on go-templates and a configuration file to template prowjob configuration across multiple branches.

Example usage (`./../k8s.io/test-infra` is the relativ path for me locally to the root of the test-infra directory):

```
TEST_INFRA_DIR=../../k8s.io/test-infra make generate-test-infra-prowjobs
```

Example output: https://github.com/kubernetes/test-infra/pull/31581

**One change to our configuration would be currently:**

Our jobs upgrading to `latest` kubernetes (currently v1.30) had `latest` in their job names. With the current implementation those will use `1-30` instead.

Example:
* Before: `periodic-cluster-api-e2e-workload-upgrade-1-29-latest-main`
* After: `periodic-cluster-api-e2e-workload-upgrade-1-29-1-30-main`

Requires to merge the following to make mdlint happy:

- https://github.com/kubernetes/test-infra/pull/31581

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #9257

/area ci

chrischdi commented:
/retest

chrischdi commented:
>      failed to get component source YAML from URL: failed to get https://github.com/kubernetes-sigs/cluster-api/releases/download/v1.6.0/core-components.yaml: got status code 500

/retest

chrischdi commented:
/retitle 🌱 [wip] hack: implement generator for prowjobs

cprivitere commented:
Any concerns that we're bringing up a library that hasn't been updated since November of 2022? (github.com/Masterminds/sprig)

chrischdi commented:
> Any concerns that we're bringing in a library that hasn't been updated since November of 2022? (github.com/Masterminds/sprig)

Great point, I did not think about that.

We already use sprig in clusterclass / in our core go.mod file which is why I think this could be considered okay.

cprivitere commented:
> > Any concerns that we're bringing in a library that hasn't been updated since November of 2022? (github.com/Masterminds/sprig)
> 
> Great point, I did not think about that.
> 
> We already use sprig in clusterclass / in our core go.mod file which is why I think this could be considered okay.

If the group already agrees it's ok, then I guess that's cool. 

Was just a bit concerned given their go.mod is based on 1.13 and they're bringing in really of versions of semver.

chrischdi commented:
To add: happy to adjust if we consider it for it should not being used, definetly not a must. Some of the sprig functions may just make it more easy to write the templates.

chrischdi commented:
Note: I removed sprig and replaced it by standard library functions.

One exception: I added a `last` function which returns the last element of an array or slice.

Also removed the other custom implemented functions.

sbueringer commented on hack/tools/prowjob-gen/config.go:
```diff
@@ -0,0 +1,56 @@
+type Config struct {
+	Branches       map[string]BranchConfig `json:"branches"`
+	Templates      []Template              `json:"templates"`
+	VersionsMapper VersionsMapper          `json:"versions"`
```
Should this be named Versions like the json key?

sbueringer commented on hack/tools/prowjob-gen/config.go:
```diff
@@ -0,0 +1,56 @@
+// BranchConfig is the branch-based configuration struct.
+type BranchConfig struct {
+	Interval                            string     `json:"interval"`
+	KubekinsImage                       string     `json:"kubekinsImage"`
```
What if we just call this Image? Are we losing anything? (I would prefer to avoid hard-coding kubekins if possible)

fabriziopandini replied:
ProwImage?

sbueringer replied:
I think this is all Prow :). I think I would go with either Image or TestImage or maybe ContainerImage (?)

fabriziopandini replied:
+1 for TestImage

sbueringer replied:
(additional context for TestImage. The container of the ProwJob Pod in which the tests are actually run is named "test")

sbueringer commented on hack/tools/prowjob-gen/generator.go:
```diff
@@ -0,0 +1,164 @@
+		return nil, err
+	}
+
+	return g, err
```
```suggestion
	return g, nil
```

sbueringer commented on hack/tools/prowjob-gen/generator.go:
```diff
@@ -0,0 +1,164 @@
+	for _, tpl := range g.config.Templates {
+		for branch := range g.config.Branches {
+			out, err := g.executeTemplate(branch, tpl.Name)
+			if err != nil && !errors.Is(err, os.ErrNotExist) {
```
Why is it okay to get a ErrNoExist error? And what happens after the if then?

chrischdi replied:
we can remove that, now that we configure the list of templates in the config 👍  great finding!

sbueringer commented on hack/tools/prowjob-gen/generator.go:
```diff
@@ -0,0 +1,164 @@
+	return nil
+}
+
+// cleanup deletes files which have the generatedFileHeader and had not been updated
```
Wondering if we should just delete all files with the header and simply do it before we write the new files

sbueringer commented on hack/tools/prowjob-gen/generator.go:
```diff
@@ -0,0 +1,164 @@
+
+// executeTemplate executes a previously parsed template with the data for a specific branch.
+func (g *generator) executeTemplate(branch, templateName string) (*bytes.Buffer, error) {
+	klog.Infof("executing template %q for branch %q", templateName, branch)
```
```suggestion
	klog.Infof("Executing template %q for branch %q", templateName, branch)
```

sbueringer commented on hack/tools/prowjob-gen/generator.go:
```diff
@@ -0,0 +1,164 @@
+	out.WriteString(generatedFileHeader)
+
+	if err := g.templates.ExecuteTemplate(&out, templateName, data); err != nil {
+		return nil, errors.Wrapf(err, "Executing template %q for branch %q", templateName, branch)
```
```suggestion
		return nil, errors.Wrapf(err, "executing template %q for branch %q", templateName, branch)
```

sbueringer commented on hack/tools/prowjob-gen/generator.go:
```diff
@@ -0,0 +1,164 @@
+	return funcs
+}
+
+func last(list any) any {
```
Can this be done with generics instead of reflection?

chrischdi replied:
I did not find a way to use generics here. May be because the template.FuncMap is a `map[string]any`

sbueringer replied:
Alright, no problem

sbueringer commented on hack/tools/prowjob-gen/main_test.go:
```diff
@@ -0,0 +1,30 @@
+)
+
+func Test_newGenerator(t *testing.T) {
+	_, err := newGenerator(Config{}, "test", "")
```
I would suggest to add "configFile" as a parameter to newGenerator and then make this case a bit more real by adding an example config file to the test/ folder.

Also makes it possible to run this test here with the debugger a bit to play around

sbueringer reviewed (commented):
First round of review. Will continue later

fabriziopandini reviewed (commented):
Overall lgtm, great to see this making progress!

sbueringer commented on hack/boilerplate/boilerplate.py:
```diff
@@ -154,7 +155,10 @@ def file_extension(filename):
-    'hack/lib/swagger.sh', 'hack/boilerplate/boilerplate.py']
+    'hack/lib/swagger.sh',
+    'hack/boilerplate/boilerplate.py',
+    '/hack/tools/prowjob-gen/generator.go',
```
The problem is that the boilerplate.py identifies every file with `DO NOT EDIT` as generated (and thus doesn't check it).

The problem is that this go file has to contain this string though. By adding generator.go here we can make sure it is checked by boilerplate.py. 

I think the name of this array is not ideal. Not sure if we want to change it though (to keep this as sync as possible with upstream)

fabriziopandini replied:
Let's just add a comment

sbueringer commented on hack/tools/prowjob-gen/README.md:
```diff
@@ -0,0 +1,68 @@
+
+  templates:
+  - name: "cluster-api-periodics.yaml.tpl"
+    format: "cluster-api-periodics-%s.yaml"
```
It would be nice if we can do something like this
```suggestion
    format: "cluster-api-periodics-{{ .branch }}.yaml"
```

This way it would be very easy to tell what gets added here. WDYT?
(not super important though)

sbueringer commented on hack/tools/prowjob-gen/README.md:
```diff
@@ -0,0 +1,68 @@
+
+When executing a template, the following variables are available:
+
+- `branch`: The branch name the file get's templated for (The key in `.prow_ignored.branches`).
```
```suggestion
- `branch`: The branch name the file gets templated for (The key in `.prow_ignored.branches`).
```

I think

sbueringer commented on .github/ISSUE_TEMPLATE/kubernetes_bump.md:
```diff
@@ -34,21 +34,25 @@ changes should be cherry-picked to all release series that will support the new
+
+  * At the `.versions`  section in the `cluster-api-prowjob-gen.yaml` file in [test-infra](https://github.com/kubernetes/test-infra/blob/master/config/jobs/kubernetes-sigs/cluster-api/):
+    * Add a new entry for the new Kubernetes version
+    * Adjust the released kKubernetes's version entry to refer `stable-1.<minor>` instead of `ci/latest-1.<minor>`
```
```suggestion
    * Adjust the released Kubernetes's version entry to refer `stable-1.<minor>` instead of `ci/latest-1.<minor>`
```

sbueringer commented on .github/ISSUE_TEMPLATE/kubernetes_bump.md:
```diff
@@ -34,21 +34,25 @@ changes should be cherry-picked to all release series that will support the new
     * Note: Only bump for Cluster API versions that will support the new Kubernetes release.
   * Prior art: #9160
```
Thx for the updates here, look great! We also have to update some docs in release-tasks.md

I think it is mostly: Setup jobs and dashboards for a new release branch

chrischdi replied:
done :-)

sbueringer commented:
Also reviewed test-infra. Really really nice work. Thank you very much!

chrischdi commented on hack/tools/prowjob-gen/config.go:
```diff
@@ -0,0 +1,56 @@
+
+// BranchConfig is the branch-based configuration struct.
+type BranchConfig struct {
+	Interval                            string     `json:"interval"`
```
*UpgradesInterval

sbueringer replied:
Yup, good point! :D

sbueringer commented:
@chrischdi No rush, but let me know when I should do another review (I'll wait until then)

chrischdi commented:
@sbueringer PTAL

pending squash

sbueringer commented:
/lgtm

/assign @fabriziopandini

sbueringer commented:
Thx!

/lgtm
/approve

/hold cancel

sbueringer commented:
/hold

sbueringer commented:
Sorry wanted to approve the test-infra PR :D 


/approve cancel

sbueringer commented:
/hold cancel
